File: quant-engine/signal_quality.py
from __future__ import annotations

from typing import Any, Dict, List, Optional


def _clamp(value: float, low: int = 0, high: int = 100) -> int:
    return int(max(low, min(high, round(value))))


def _confidence_level(confidence: int) -> str:
    if confidence >= 70:
        return "HIGH"
    if confidence >= 50:
        return "MEDIUM"
    return "LOW"


def evaluate_signal_quality(
    symbol: str,
    market: str = "US",
    selected_provider: Optional[str] = None,
    market_data_quality: Optional[Dict[str, Any]] = None,
    technical_indicators: Optional[Dict[str, Any]] = None,
    ml_prediction: Any = None,
    risk_metrics: Optional[Dict[str, Any]] = None,
    graham_result: Optional[Dict[str, Any]] = None,
    sentiment_result: Optional[Dict[str, Any]] = None,
    workflow_action: str = "HOLD",
    workflow_confidence: int = 0,
    reasons: Optional[List[str]] = None,
) -> Dict[str, Any]:
    quality = market_data_quality or {}
    technical = technical_indicators or {}
    risk = risk_metrics or {}
    graham = graham_result or {}
    sentiment = sentiment_result or {}
    reasons = reasons or []

    supporting: List[str] = []
    contradicting: List[str] = []
    blocking: List[str] = []
    warnings: List[str] = []

    quality_score = int(quality.get("quality_score") or 0)
    usable_for_ml = bool(quality.get("usable_for_ml"))
    usable_for_chart = bool(quality.get("usable_for_chart"))

    if not usable_for_ml:
        blocking.append("market_data_quality.usable_for_ml=false")
    if quality_score < 60:
        blocking.append(f"market_data_quality.quality_score={quality_score} < 60")
    if usable_for_chart and not usable_for_ml:
        warnings.append("Datos aptos solo para grafico; BUY/SELL prohibido.")

    action = str(workflow_action or "HOLD").upper()
    confidence = _clamp(int(workflow_confidence or 0))
    signal_score = confidence

    ml_value = ml_prediction.get("expected_return") if isinstance(ml_prediction, dict) else ml_prediction
    try:
        ml_value = float(ml_value or 0)
    except Exception:
        ml_value = 0.0

    var_95 = risk.get("var_95", risk.get("var_1d_95", risk.get("var")))
    try:
        var_95 = float(var_95 if var_95 is not None else 0)
    except Exception:
        var_95 = 0.0
    high_risk = var_95 >= 0.05 or bool(risk.get("high_risk"))

    sentiment_label = str(sentiment.get("sentiment") or sentiment.get("label") or "NEUTRAL").upper()
    graham_passed = graham.get("passed", graham.get("graham_passed"))
    graham_reason = str(graham.get("reason") or graham.get("graham_reason") or "")

    if ml_value > 0.01:
        supporting.append("ML positivo")
        signal_score += 10
    elif ml_value < -0.01:
        contradicting.append("ML negativo")
        signal_score -= 10

    if high_risk:
        contradicting.append(f"Riesgo alto VaR={var_95}")
        signal_score -= 25
    elif var_95 > 0:
        supporting.append(f"Riesgo controlado VaR={var_95}")
        signal_score += 5

    if graham_passed is False:
        contradicting.append(f"Graham negativo: {graham_reason or 'sin margen de seguridad'}")
        signal_score -= 15
    elif graham_passed is True:
        supporting.append("Graham positivo")
        signal_score += 5

    if sentiment_label == "POSITIVE":
        if blocking:
            blocking.append("FinBERT positivo ignorado por mala calidad de datos")
        elif ml_value <= 0 and action != "BUY":
            warnings.append("FinBERT positivo no confirma tecnico/ML; no infla la senal.")
        else:
            supporting.append("FinBERT positivo")
            signal_score += 5
    elif sentiment_label == "NEGATIVE":
        contradicting.append("FinBERT negativo")
        signal_score -= 10

    if technical.get("macd_signal") in {"Cruce alcista", "Positivo"} or technical.get("momentum") == "bullish":
        supporting.append("Tecnico alcista")
    if technical.get("macd_signal") == "Cruce bajista" or technical.get("momentum") == "bearish":
        contradicting.append("Tecnico bajista")

    if action == "BUY" and high_risk:
        contradicting.append("Workflow BUY contradicho por riesgo alto")
    if action == "BUY" and graham_passed is False:
        contradicting.append("Workflow BUY contradicho por Graham negativo")

    if blocking:
        final_action = "HOLD"
        final_confidence = 0
        signal_status = "BLOCKED"
        signal_score = 0
    else:
        signal_score = _clamp(signal_score)
        if action == "BUY" and (high_risk or graham_passed is False or len(contradicting) >= 2):
            final_action = "HOLD"
            final_confidence = min(confidence, 49)
            signal_status = "CONFLICTED"
        elif action in {"BUY", "SELL"} and confidence >= 70 and supporting:
            final_action = action
            final_confidence = _clamp(signal_score)
            signal_status = "OK"
        elif action in {"BUY", "SELL"} and confidence >= 50:
            final_action = "HOLD" if contradicting else action
            final_confidence = min(_clamp(signal_score), 69)
            signal_status = "CONFLICTED" if contradicting else "WEAK"
        else:
            final_action = "HOLD"
            final_confidence = min(_clamp(signal_score), 49)
            signal_status = "WEAK"

    if not supporting and not blocking:
        warnings.append("Sin factores de soporte suficientes para oportunidad.")

    explanation = (
        f"{signal_status}: accion final {final_action} con confianza {final_confidence}. "
        f"Soporte={len(supporting)}, contradicciones={len(contradicting)}, bloqueos={len(blocking)}."
    )

    return {
        "symbol": symbol,
        "signal_status": signal_status,
        "final_action": final_action,
        "final_confidence": final_confidence,
        "confidence_level": _confidence_level(final_confidence),
        "signal_score": _clamp(signal_score),
        "supporting_factors": supporting,
        "contradicting_factors": contradicting,
        "blocking_reasons": blocking,
        "warnings": warnings,
        "explanation": explanation,
        "raw_diagnostics": {
            "market": market,
            "selected_provider": selected_provider,
            "workflow_action": action,
            "workflow_confidence": confidence,
            "quality_score": quality_score,
            "usable_for_ml": usable_for_ml,
            "ml_prediction": ml_value,
            "var_95": var_95,
            "graham_passed": graham_passed,
            "sentiment": sentiment_label,
            "reasons": reasons,
        },
    }

File: src/lib/signal-quality.ts
import type { MarketDataQualityResult } from './market-data-quality'

export type SignalStatus = 'OK' | 'WEAK' | 'CONFLICTED' | 'BLOCKED'
export type FinalAction = 'BUY' | 'SELL' | 'HOLD'
export type ConfidenceLevel = 'LOW' | 'MEDIUM' | 'HIGH'

export interface SignalQualityResult {
  symbol: string
  signal_status: SignalStatus
  final_action: FinalAction
  final_confidence: number
  confidence_level: ConfidenceLevel
  signal_score: number
  supporting_factors: string[]
  contradicting_factors: string[]
  blocking_reasons: string[]
  warnings: string[]
  explanation: string
  raw_diagnostics: Record<string, unknown>
}

function clamp(value: number) {
  return Math.max(0, Math.min(100, Math.round(value)))
}

function confidenceLevel(confidence: number): ConfidenceLevel {
  if (confidence >= 70) return 'HIGH'
  if (confidence >= 50) return 'MEDIUM'
  return 'LOW'
}

export function assessSignalQuality(input: {
  symbol: string
  market: string
  selected_provider?: string | null
  market_data_quality?: MarketDataQualityResult
  technical_indicators?: Record<string, unknown>
  ml_prediction?: unknown
  risk_metrics?: Record<string, unknown>
  graham_result?: Record<string, unknown>
  sentiment_result?: Record<string, unknown>
  workflow_action?: string
  workflow_confidence?: number
  reasons?: string[]
}): SignalQualityResult {
  const quality = input.market_data_quality
  const technical = input.technical_indicators ?? {}
  const risk = input.risk_metrics ?? {}
  const graham = input.graham_result ?? {}
  const sentiment = input.sentiment_result ?? {}
  const supporting: string[] = []
  const contradicting: string[] = []
  const blocking: string[] = []
  const warnings: string[] = []

  const qualityScore = Number(quality?.quality_score ?? 0)
  if (!quality?.usable_for_ml) blocking.push('market_data_quality.usable_for_ml=false')
  if (qualityScore < 60) blocking.push(`market_data_quality.quality_score=${qualityScore} < 60`)
  if (quality?.usable_for_chart && !quality?.usable_for_ml) warnings.push('Datos aptos solo para grafico; BUY/SELL prohibido.')

  const action = String(input.workflow_action || 'HOLD').toUpperCase()
  const confidence = clamp(Number(input.workflow_confidence ?? 0))
  let score = confidence
  const mlValue = Number(typeof input.ml_prediction === 'object' && input.ml_prediction !== null
    ? (input.ml_prediction as Record<string, unknown>).expected_return
    : input.ml_prediction) || 0
  const var95 = Number(risk.var_95 ?? risk.var_1d_95 ?? risk.var ?? 0) || 0
  const highRisk = var95 >= 0.05 || risk.high_risk === true
  const grahamPassed = graham.passed ?? graham.graham_passed
  const sentimentLabel = String(sentiment.sentiment ?? sentiment.label ?? 'NEUTRAL').toUpperCase()

  if (mlValue > 0.01) { supporting.push('ML positivo'); score += 10 }
  if (mlValue < -0.01) { contradicting.push('ML negativo'); score -= 10 }
  if (highRisk) { contradicting.push(`Riesgo alto VaR=${var95}`); score -= 25 }
  else if (var95 > 0) { supporting.push(`Riesgo controlado VaR=${var95}`); score += 5 }
  if (grahamPassed === false) { contradicting.push('Graham negativo'); score -= 15 }
  if (grahamPassed === true) { supporting.push('Graham positivo'); score += 5 }

  if (sentimentLabel === 'POSITIVE') {
    if (blocking.length) blocking.push('FinBERT positivo ignorado por mala calidad de datos')
    else if (mlValue <= 0 && action !== 'BUY') warnings.push('FinBERT positivo no confirma tecnico/ML; no infla la senal.')
    else { supporting.push('FinBERT positivo'); score += 5 }
  }
  if (sentimentLabel === 'NEGATIVE') { contradicting.push('FinBERT negativo'); score -= 10 }

  if (technical.macd_signal === 'Cruce alcista' || technical.macd_signal === 'Positivo') supporting.push('Tecnico alcista')
  if (technical.macd_signal === 'Cruce bajista') contradicting.push('Tecnico bajista')
  if (action === 'BUY' && highRisk) contradicting.push('Workflow BUY contradicho por riesgo alto')
  if (action === 'BUY' && grahamPassed === false) contradicting.push('Workflow BUY contradicho por Graham negativo')

  let final_action: FinalAction = 'HOLD'
  let final_confidence = 0
  let signal_status: SignalStatus = 'WEAK'

  if (blocking.length) {
    signal_status = 'BLOCKED'
    score = 0
  } else {
    score = clamp(score)
    if (action === 'BUY' && (highRisk || grahamPassed === false || contradicting.length >= 2)) {
      signal_status = 'CONFLICTED'
      final_confidence = Math.min(confidence, 49)
    } else if ((action === 'BUY' || action === 'SELL') && confidence >= 70 && supporting.length) {
      final_action = action as FinalAction
      final_confidence = clamp(score)
      signal_status = 'OK'
    } else if ((action === 'BUY' || action === 'SELL') && confidence >= 50) {
      final_action = contradicting.length ? 'HOLD' : action as FinalAction
      final_confidence = Math.min(clamp(score), 69)
      signal_status = contradicting.length ? 'CONFLICTED' : 'WEAK'
    } else {
      final_confidence = Math.min(clamp(score), 49)
    }
  }

  if (!supporting.length && !blocking.length) warnings.push('Sin factores de soporte suficientes para oportunidad.')
  if (signal_status === 'BLOCKED') final_confidence = 0

  return {
    symbol: input.symbol,
    signal_status,
    final_action,
    final_confidence,
    confidence_level: confidenceLevel(final_confidence),
    signal_score: clamp(score),
    supporting_factors: supporting,
    contradicting_factors: contradicting,
    blocking_reasons: blocking,
    warnings,
    explanation: `${signal_status}: accion final ${final_action} con confianza ${final_confidence}. Soporte=${supporting.length}, contradicciones=${contradicting.length}, bloqueos=${blocking.length}.`,
    raw_diagnostics: {
      market: input.market,
      selected_provider: input.selected_provider,
      workflow_action: action,
      workflow_confidence: confidence,
      quality_score: qualityScore,
      usable_for_ml: quality?.usable_for_ml === true,
      ml_prediction: mlValue,
      var_95: var95,
      sentiment: sentimentLabel,
      reasons: input.reasons ?? [],
    },
  }
}
